homeassistant/components/remootio/utils.py

```diff
@@ -0,0 +1,114 @@
+"""Utility methods for the Remootio integration."""
+import logging
+from logging import Logger
+
+from aioremootio import ConnectionOptions, LoggerConfiguration, RemootioClient
+from aioremootio.enums import State
+from aioremootio.errors import (
+    RemootioClientAuthenticationError,
+    RemootioClientConnectionEstablishmentError,
+)
+import async_timeout
+
+from homeassistant import core
+from homeassistant.helpers import aiohttp_client
+
+from .const import EXPECTED_MINIMUM_API_VERSION, REMOOTIO_TIMEOUT
+from .exceptions import (
+    CannotConnect,
+    InvalidAuth,
+    UnsupportedRemootioApiVersionError,
+    UnsupportedRemootioDeviceError,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def _check_api_version(remootio_client: RemootioClient) -> None:
+    """Check whether the by the given client represented Remootio device uses a supported API version."""
+    api_version: int = await remootio_client.api_version
+    if api_version < EXPECTED_MINIMUM_API_VERSION:
+        raise UnsupportedRemootioApiVersionError(remootio_client, api_version)
+
+
+async def _check_sensor_installed(
+    remootio_client: RemootioClient, raise_error: bool = True
+) -> None:
+    """Check whether the by the given client represented Remootio device has a sensor installed."""
+    if await remootio_client.initialized:
```

Comment:
Can't really await a property? Futures are also not safe to await from multiple tasks if this happen to be a future.

Reply:
> Can't really await a property? Futures are also not safe to await from multiple tasks if this happen to be a future.

When instantiating the `RemootioClient` class, a lot of steps happens:
1. a connection is established to the device;
2. the [authentication](https://github.com/remootio/remootio-api-documentation/blob/master/websocket_api_v3_specification.md#7-auth-frame-and-the-authentication-flow) takes place (whereby the current status of the device is also found out);
3. a [HELLO frame](https://github.com/remootio/remootio-api-documentation/blob/master/websocket_api_v3_specification.md#hello-frame) is sent to the device and the response is awaited to find out the serial number of the device and the API version used by the device;
4. two asynchronous tasks are started to send [PING frames](https://github.com/remootio/remootio-api-documentation/blob/master/websocket_api_v3_specification.md#ping-frame) to the device and to receive and handle frames from the device.

The whole of the above initialization process takes place partly asynchronously and is encompassed by a [`Lock`](https://docs.python.org/3/library/asyncio-sync.html#asyncio.Lock). When the (asynchronous) property `initialized` is queried, it is queried whether this `Lock` is active, and if so, the unlocking of it is awaited and only then the value of the property is returned.

Reply:
I understand that. I just don't think this should be a property. It looks confusing. I would just have expected a function called initialize() which was awaited right after construction of the object. But it's a minor detail.

Reply:
I don't want to rewrite the whole library now.

Reply:
Not really a rewrite to convert a property to a function. Bit as said, it's a minor detail. Can remain.